treewide: Add wrapper for `netlink` functions that may fail with `ErrDumpInterrupted`

gandro opened the pull request:
According to the kernel docs[^1], the kernel can return incomplete results for netlink state dumps if the state changes while we are dumping it. The result is then marked by `NLM_F_DUMP_INTR`. The `vishvananda/netlink` library returned `EINTR` since v1.2.1, but more recent versions have changed it such that it returns `netlink.ErrDumpInterrupted` instead[^2].

These interruptions seem common in high-churn environments. If the error occurs, it is in most cases best to just try again.  Therefore, this commit adds a wrapper for all `netlink` functions marked to return `ErrDumpInterrupted` that retries the function up to 30 times until it either succeeds or returns a different error.

While may call sites do have their own high-level retry mechanism (see e.g. https://github.com/cilium/cilium/pull/32099), the logged error message can still cause CI to fail (e.g. https://github.com/cilium/cilium/pull/35259). Long high-level retry intervals can also become problematic: For example, if the routing setup fails due to `NLM_F_DUMP_INTR` during an CNI ADD invocation, the retry adds add several seconds of additional delay to an already overloaded system, instead of resolving the issue quickly.

The last commit adds an additional linter that nudges developers to use this new `safenetlink` package for function calls that can be interrupted. This ensures that we don't have to add retries in all subsystems individually.

**Review per commit**

[^1]: https://docs.kernel.org/userspace-api/netlink/intro.html
[^2]: https://github.com/vishvananda/netlink/pull/1018

gandro committed:
netlink: Add wrapper for functions that fail with `ErrDumpInterrupted`

According to the kernel docs[^1], the kernel can return incomplete
results for netlink state dumps if the state changes while we are
dumping it. The result is then marked by `NLM_F_DUMP_INTR`. The
`vishvananda/netlink` library returned `EINTR` since v1.2.1, but more
recent versions have changed it such that it returns
`netlink.ErrDumpInterrupted` instead[^2].

These interruptions seem common in high-churn environments. If the error
occurs, it is in most cases best to just try again.  Therefore, this
commit adds a wrapper for all `netlink` functions marked to return
`ErrDumpInterrupted` that retries the function up to 30 times until it
either succeeds or returns a different error.

While may call sites do have their own high-level retry mechanism (see
e.g. cilium/cilium#32099), the logged error message can still cause CI
to fail (e.g. cilium/cilium#35259). Long high-level retry intervals can
also become problematic: For example, if the routing setup fails due to
`NLM_F_DUMP_INTR` during an CNI ADD invocation, the retry adds add
several seconds of additional delay to an already overloaded system,
instead of resolving the issue quickly.

A subsequent commit will add an additional linter that nudges developers
to use this new `safenetlink` package for function calls that can be
interrupted. This ensures that we don't have to add retries in all
subsystems individually.

[^1]: https://docs.kernel.org/userspace-api/netlink/intro.html
[^2]: https://github.com/vishvananda/netlink/pull/1018

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
bugtool: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
`ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
cilium-dbg: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
`ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
daemon: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
cilium-cni: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
cilium-health: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
wireguard: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
pkg/node: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
proxy: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
mac: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
mtu: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
datapath: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

In addition, this commit also removes some existing retry mechanisms with the
canoncial safenetlink version.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
endpoint: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
ipam: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
multicast: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
contrib: Add linter to ensure usage of safenetlink

This adds a new check script to ensure any functions which has a safe
wrapper in the `safenetlink` package uses the safe variant.

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro commented on pkg/datapath/linux/config/config.go:
```diff
@@ -895,7 +878,7 @@ func vlanFilterMacros(nativeDevices []*tables.Device) (string, error) {
 	for _, v := range vlansByIfIndex {
 		vlansCount += len(v)
-		slices.Sort(v) // sort Vlanids in-place since netlink.LinkList() may return them in any order
+		slices.Sort(v) // sort Vlanids in-place since safenetlink.LinkList() may return them in any order
```
Note to reviewers: This is necessary so `check-safenetlink.sh` doesn't complain about it.

bimmlerd commented on CODEOWNERS:
```diff
@@ -594,6 +594,7 @@ Makefile* @cilium/build
 /pkg/resiliency @cilium/sig-agent
 /pkg/revert/ @cilium/sig-agent
 /pkg/safeio @cilium/sig-agent
+/pkg/safenetlink @cilium/sig-datapath
```
netlink afaik is Linux (& FreeBSD) specific, could have it under `pkg/datapath/linux/safenetlink`?

gandro replied:
So it is imported by the non-Linux builds of Cilium CLI (which is why I had to add the non-Linux stubs). But yeah, it is overall very Linux specific, we could move it.

gandro replied:
Done

bimmlerd reviewed (approved):
codeowners and custom vet seem sensible. Did you also look for usages of the `netlink/nl` pkg?

derailed reviewed (approved):
@gandro Nice work Sebastian! Looks good pending the lint resolutiuon.

joestringer commented on pkg/safenetlink/netlink_linux.go:
```diff
@@ -0,0 +1,395 @@
+)
+
+// withRetry runs the netlinkFunc. If netlinkFunc returns netlink.ErrDumpInterrupted, the function is retried up
+// to 20 times. If success or any other error is returned, withRetry returns immediately, propagating the error.
```
nit: The constant is already wrong, drop it to be consistent
```suggestion
// withRetry runs the netlinkFunc. If netlinkFunc returns netlink.ErrDumpInterrupted, the function is retried. If success or any other error is returned, withRetry returns immediately, propagating the error.
```

joestringer commented on pkg/safenetlink/netlink_linux.go:
```diff
@@ -0,0 +1,395 @@
+
+// withRetry runs the netlinkFunc. If netlinkFunc returns netlink.ErrDumpInterrupted, the function is retried up
+// to 20 times. If success or any other error is returned, withRetry returns immediately, propagating the error.
+func withRetry(netlinkFunc func() error) error {
```
Should we have metrics or some other observability tooling to introspect netlink commands that are hitting high churn, maybe count when the number of retries is high or something? I'm thinking that if Cilium is actively hitting an issue with one of these commands that causes either undue delay in configuring the kernel / dumping state or such then this would be nice to know as a hint if we experience any issues with some subsystem in a production environment.

gandro replied:
>  The main question to me would be whether this could cause significant delay in executing some netlink command deep in some callstack, maybe leading to other problems in this adverse case. On that front I raised a question about metrics below.

With the current parameters, the maximum delay is 50 milliseconds. After that, the call unblocks and the wrapper returns an error.

Some of the code I've removed and replaced with the wrapper actually busy-looped forever.

> Should we have metrics or some other observability tooling to introspect netlink commands that are hitting high churn, maybe count when the number of retries is high or something? 

I'll see if that is easily doable. One issue is that the wrappers are also used by cilium-cni (which is actually the main motivation for this PR, since we observed issues with CNI ADD) and in those wrappers the metrics won't work.

gandro replied:
I attempted to add metrics, but it ends up being kinda ugly and complicated due to 

a) the fact that I want the `safenetlink` to act as a drop-in replacement for the `netlink` calls and thus having to use global variables to store the counters
b) the fact that the existing `netlink` is transitively used by the `metrics` package (via `pkg/option`). Thus, defining metrics in the `safenetlink` package introduces a circular dependency :dizzy_face: Breaking this cycling dependency requires some shuffling of code that I'd like to avoid doing in this PR (that we should backport)

joestringer reviewed (approved):
I didn't specifically analyse whether any of my codeowners areas already have retries, but overall this seems safe enough anyway. The main question to me would be whether this could cause significant delay in executing some netlink command deep in some callstack, maybe leading to other problems in this adverse case. On that front I raised a question about metrics below.

Non-blocking feedback follows.

squeed commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -61,7 +62,7 @@ func (f *GenericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.Plug
 
 	if err = ns.Do(func() error {
-		links, err := netlink.LinkList()
+		links, err := safenetlink.LinkList()
```
FWIW (I'll look in to this) I think we can get rid of this LinkList call anyways. It should always be the interface specified by CNI_IFNAME.

gandro committed:
l2responder: Use safenetlink package

See commit `netlink: Add wrapper for functions that fail with
ErrDumpInterrupted` for details.

This commit is a bit different from the previous ones as it uses the
retry helper rather than a wrapper.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
